designs/CentralNuGetManagement/CPVM-Project_Package_Version_Override.md

```diff
@@ -0,0 +1,425 @@
+# CPVM Project Package Version Override
+
+Users have indicated that Central Package Version Management (CPVM) needs a mechanism to override the centrally managed versions of packages on a project-by-project basis. Project Package Version Override allows individual projects to import different versions of NuGet packages organized in the same, centralized location.
+
+# Design Goals
+
+1. Even though package versions need to be specified for specific projects, keep the package versions centralized. Avoid situations where solution-level and project-level package version management are mixed in the same repository.
+2. Provide a means for organizing multiple package versions together.
+3. Manage the package versions of transitive dependencies alongside direct dependencies.
+4. Follow the MSBuild approach with minimal changes to existing XML patterns.
+
+# Feature Behavior
+
+- All `<PackageVersion>` should be defined within `Directory.Packages.props`, and `Version` may not be defined on `<PackageReference>` within projects.
+    - Specifying a `Version` on a `<PackageReference>` will throw a build error.
+    - Specifying a `<PackageVersion>` within a project file is an anti-pattern, and it will generate a build time warning. It will not throw an error, and a developer may choose to suppress the warning by specifying `<AllowProjectPackageVersion>true</AllowProjectPackageVersion>` within a `<PropertyGroup>`.
+- A `<PackageVersion>` may have a `CentralManagementGroup` defined which allows a `<PackageReference>` to refer to this version using the `CentralManagementGroup` attribute.
+    - When a `Version` to be imported by a `<PackageReference>` needs to be different than the global group defined in `Directory.Packages.props`, a `CentralManagementGroup` can be specified on both the `<PackageReference>` in a project file and a corresponding `<PackageVersion>` in `Directory.Packages.props` that specifies the needed `Version`.
+    - A `<PackageReference>` may only have one `CentralManagementGroup` defined.
+    - A project can include references to multiple `CentralManagementGroup`.
+    - When `CentralManagementGroup` is not specified, both `<PackageVersion>` and `<PackageReference>` reference a global `CentralManagementGroup` that is automatically created.
+- Transitive dependencies are resolved based on finding a common level of specificity between direct imports.
```

Comment:
Have we established that folks need to refer to multiple groups for their packages in a project? How rigid is this requirement? The reason I ask this is because of the resolution use-cases explained below. It makes the override feature too complicated and the package resolution logic even more. Try explaining to customers before they pull their hair 😄.
What does `paket` do here? Can we sync up with them to understand their learnings in this space? I can provide contacts.

Reply:
The transitive application approach sounds very, very complicated. Both from a technical + UX perspective. 

Predictability is one of the things that NuGet's dependency resolution strives for (assuming you don't use floating versions), and the transitive dependency version inference additions are pretty complex. 

I think my perspective is largely naive, but I can see myself wanting to change a version or two. Mixing many `groups` creates a difficult mental model.